### ISSUE TITLE
Individual filter for each project

### ISSUE THREAD

**Issue by mitxel-m** [Maproom/qmapshack#43]
The new search on workspace feature is excellent...Thanks to pingurus and all the team ;-)

What about adding a filter for projects with the same procedure?

**Use case:**

You want to filter tracks below 1500 m from your 'Skitour' folder, you can use the search on workspace: elevation < 1500 m, all items from all projects out of the search criteria disappear. 

Ok you have found your tracks. Now you want to work with this filtered set, but at the same time you need to see on the map other items like your Summit wp collection, or other tracks from other projects/folders. For now the workaround is to copy the filtered set to a new temp project and then clear the filter on workspace. This is not efficient. 

 It would be nice to have a 'Filter Project' feature that would work only for the related Project (in addition to actual Search on workspace)

**How would it work**

Right click on a Project - select  'Filter...'

![menu](https://i.ibb.co/bHPbgC5/Filter-menu.jpg)

A dialog or a box appears to type the filter

![box2](https://i.ibb.co/hYVR7wt/search-box2.jpg)

Close or accept 

Now you have filtered items from this project besides all the elements from other projects. 

...and the Project appears marked with a Filter icon in the workspace list:

![icon](https://i.ibb.co/cyHvB35/Filter-icon-prj.jpg)

**How to clear the filter**

Right click on a Project - select  'Filter...'- click on the X 

![box2](https://i.ibb.co/hYVR7wt/search-box2.jpg)

(another approach: in the menu add an action 'Clear Filter', faster but enlarges the menu... I doubt which is better)

**Other considerations**
- The Project Filter is exclusive for the related Project, so you could apply at the same time a filter for every Project with different criteria. 
- You could apply at the same time the Workspace filter and a Project Filter, in this case you would get from the filtered project items matching both criteria.

**Comment by kiozen** [comment]
Sounds a bit like another step to abusing the workspace as database replacement. The **idea is not** to have all your data in the workspace and to filter out the stuff you don't need. The **idea is** to have only the data you are working on in the workspace.

**Comment by pingurus** [comment]
> Sounds a bit like another step to abusing the workspace as database replacement. The idea is not to have all your data in the workspace and to filter out the stuff you don't need. The idea is to have only the data you are working on in the workspace.

Then it would be good if the filter could be used in the database window aswell, so one can filter before loading. Or, the easier (from a coding perspective) solution to make people use it the way intended would be to rename the "delete" menu entry in the workspace to "remove from workspace", which would make clear(er), that the items then are not deleted from the database but just hidden and that they can be ticked and taken back to the workspace anytime. This then would of course require inverse logic from the user to first filter for the items not wanted and remove those really unwanted before then clearing the filter and having everything wanted.

**Comment by kiozen** [comment]
You can't apply that kind of search in the database as there is only a very reduced dataset available. This is not a full featured GIS database. 

* The intended use of the database is to store items and to avoid copies by referencing them. 
* The intended use of the workspace is to display and change items. 
* The intended use of the filter is to temporarily reduce displayed items. 

I know there is a "he who dies with the most wins" fraction that has huge collections of waypoints and stuff loaded permanently on the workspace. This is always a bit of a problem. I use some of these collections, too. Usually I load them, select the items of interest (area selection), copy them into my active project and unload the collection.

**Comment by pingurus** [comment]
I see. 
So what do you think of renaming the "delete" menu entry independetly of wether we implement this feature request?
The feature request itself should not be too hard to implement, if I have the code correctly in my mind, so it would rather be a design decision if we implement this or not.

**Comment by kiozen** [comment]
Yes `delete` is a bit misleading. `remove` should be more accurate. `remove from workspace` is a bit long and as we are in the workspace it should be obvious. Any idea for a better icon? The waste bin doesn't seem to fit `remove`

I am still not decided if I am happy about these filters. Another question to discuss will be the persistence of the filter when restarting QMS. I would say it's not persistent. If it should be persistent it has to be stored in the binary version of the project. And by that it would be stored into the database, too.

**Comment by pingurus** [comment]
> Yes `delete `is a bit misleading. `remove `should be more accurate. `remove from workspace` is a bit long and as we are in the workspace it should be obvious. Any idea for a better icon? The waste bin doesn't seem to fit remove

How about `hide` and the symbol that is already used for closing projects?

**Comment by kiozen** [comment]
No it's not `hide`. And on second thought it's more complex. It's kind of `unload` for database projects. And a real `remove`/`delete` for file based projects.

**Comment by mitxel-m** [comment]
Wow, I didn't expect such a fervent discussion.


may be it sounds like a bit like another step to abusing the workspace as database replacement.... but IMHO it is not.

This feature is independent of whether I use GPX or database entries. Please note that in the example I have used precisely one entry from my database, and I have intentionally put other entries in gpx format to indicate that it would work in both cases. 

And the most important, it also is independent of how many entries I have loaded into the workspace. It is the same proposal if I only have three projects or database folders than if I have 10.  

I need to filter the items of one project/folder so the others don't disappear. 

And I aspire not to duplicate them in my active project, or in another temporary one because I only need the elements to visualize them while I plan another track. (use case: you want to draw a new track based on the filtered ones)

- if I import them into my work project I can't handle them as a layer showing/hiding them independently of what I'm drawing, 
- If I am forced to copy them to a temporary project then I am increasing workspace.db unnecessarily.

 And seriously, if I have been able to load them without problem before filter them,  what is the problem in keeping them filtered as a layer while I work on another project?

QMS handles these kind of collections very well, even when they are large.  And it is a great advantage to work with them as layers that can be shown/hidden in the canvas with a click.

But, of course, once the work is finished, it is logical to close that entries in order to keep the workspace with as few elements as possible.

@Kiozen Please, note that my proposal does not exclude or limit your current way of doing things.

**Comment by mitxel-m** [comment]
regarding the parallel discussion about delete/remove/hide.

I find appropriate the term `delete` . Maybe `remove` is more accuratey, but `delete` also describes the action perfectly : you are deleting the element **from the database entry**, not from the database itself.

In the wiki it is perfectly explained with those terms and I think it is clear.

**Comment by pingurus** [comment]
> you are deleting the element from the database entry, not from the database itself.

No, what confused me for a time was, that you don't remove it from the project, but only unload it. When you close and then reopen the project, it is in the project again. So in my mind this is no deletion but only a temporary action.
But maybe I did not get yet how the QMS database system is intended to be used :smile: 
I think I will make some kind of graph how I understand the system and then, if it is the way it works, add it to the wiki.

**Comment by pingurus** [comment]
> @kiozen Please, note that my proposal does not exclude or limit your current way of doing things.

Pros:
* Strongly improved usability

Cons:
* Database design choice (however I personally don't think that it is that big of a "step to abusing the workspace as database replacement", if any)
* longer context menu
* "You  could apply at the same time the Workspace filter and a Project Filter,  in this case you would get from the filtered project items matching  both criteria." This won't work with the current implementation without redesigning the interfaces. (But see below)

Thus, I would happily try to make it work. The cooperation of both searches might well be an easy to implement sideffect of the actual solution to make the `project filtered` symbol show at the right times.

**Comment by kiozen** [comment]
If you `delete` an item from a database project in the workspace it's not deleted in the database. It's simply removed from the workspace. If you save the project just the items in the workspace are saved. All other items of the project are not affected.

However if you delete an item from a normal file based project it's gone and if you save that project the file is completely written from scratch loosing all information about the deleted file. So this is a real delete.

We can change this dynamically in the menu. But this is another issue on it's own.

**Comment by kiozen** [comment]
> I would happily try to make it work. The cooperation of both searches might well be an easy to implement sideffect of the actual solution to make the `project filtered` symbol show at the right times.

That's what I usually do with issues I am not really happy about. I try to implement. If it turns out to be a nothing - fine. If it turns out to be a larger effort I drop it. No sense in making the code more complicated for features that don't fit nicely.

**Comment by mitxel-m** [comment]
> _"You could apply at the same time the Workspace filter and a Project Filter, in this case you would get from the filtered project items matching both criteria."_   
>This won't work with the current implementation without redesigning the interfaces. 

Ok, I have only pointed that out to warn that two filters could come together, and what "could" be the result.

But it is not essential to obtain results by the two criteria, and if that point complicates things, a solution could be to prioritize the workspace search over the project filter

I mean, while there is a workspace search, the project filter would be overridden until the workspace search is cleared.

This makes sense because now also when you do a search all the items that meet the criteria are shown in the workspace. Even the item list expands in those projects that are collapsed and/or not marked for show items on the map. Therefore the same behaviour would be understandable for projects that have a filter.

To indicate that a project has its filter temporarily overridden by the workspace search it could lose its filter icon, or get it changed to another one like this:

![FilterOver.png](https://i.ibb.co/zNQWMPs/Filter-Over.jpg)

**Comment by pingurus** [comment]
> I mean, while there is a workspace search, the project filter would be overridden until the workspace search is cleared.

That would be more complicated porbably :smile:
I'm nearly done with implementing a first version where both filters are combined, I only need the workspace to show that a filter is applied. (simply an issue of me understanding completely how the workspace tree works :smile: )

**Comment by pingurus** [comment]
I now need input of how you would like to enter the query, in the pull request I'm about to make it is simply an input dialog, I don't find that too nice. Also there should be a better way to remove the query than opening the input window, removing the text and then pressing ok.

**Comment by kiozen** [comment]
you should make the input line a widget on it's own to be used on top of the workspace tree widget as well as floating widget when triggered by a menu.

And probably it's a good idea to reset the filter when pressing the filter icon or another icon next to it.

**Comment by pingurus** [comment]
> you should make the input line a widget on it's own to be used on top of the workspace tree widget as well as floating widget when triggered by a menu.

Have a look if the PR is what you thought of.

> And probably it's a good idea to reset the filter when pressing the filter icon or another icon next to it.

Shouldn't there appear the same icon as in the workspace search? I'm not sure though where.

**Comment by mitxel-m** [comment]
Downloaded and tested. The filters work Ok :-) 

> I'm about to make it is simply an input dialog, I don't find that too nice. Also there should be a better way to remove the query than opening the input window, removing the text and then pressing ok.

I agree. Maybe It would be better  a box ( a clone of the workspace search box)  in the first line of the project tree. This would only appear if you trigger the filter action from the context menu, and disappear when clearing the filter string with the X.

I am preparing a screenshot and a notes with some hints...

**Comment by mitxel-m** [comment]
After testing the PR, some hints and a bit of screenshot art :-)

Placing the filter icons directly in the header line of the project (as I initially proposed) brings some drawbacks:

- The project name has its own tooltip that gives some information about where it lives and the content. By placing the filter icons in the same line that tooltip is overridden by the filter icons tooltips

![tiptool-prj.jpg](https://i.ibb.co/T8m803C/tiptool-prj.jpg)

- If you have marked a project as `Active project` it is advised by a star icon. If the user also applies a project filter (not logical but possible) that icon is overridden by the filter icon, and worst of all, when you remove the filter it is not restored.

![active-prj.jpg](https://i.ibb.co/ncDZVLf/active-prj.jpg)

So it would be a good idea to leave the project header line as it is, and try to put a search box just below, in the first line of the project tree, as Oliver suggested.

It would be a clone of the workspace search box. Something like this:

![Filter-prj-box-1.jpg](https://i.ibb.co/ggY3gvm/Filter-prj-box-1.jpg) 


![Filter-prj-box-2.jpg](https://i.ibb.co/QDRcVsZ/Filter-prj-box-2.jpg)

doing so have some pros:
- It follows the logic of other search boxes along the interface (workspace search, Maps and DEM search,...), and therefore is more coherent and intuitive to use
- The help and warning icons can be used inside the box right as in the workspace search.
- keeping the box in view while the filter is active makes the user aware of which is the query
- It is also practical to edit the query without needing to go through the context menu again. 

It would be ideal if this box only appears when called from the ProjectFilter action in the context menu, maybe it could be hidden when you reset the filter by pressing the X, but...

A better aproach might be to place "Enable filter..." in the context menu as a check or toggle: 
- when you  check/press it -> the action would be show the search box in the first line of the project tree 
- when you uncheck/unpress it the action would be clear the query and hide the box.

All this said from a user point of view,  (I hope it also makes sense for the code.)
;-)

**Comment by mitxel-m** [comment]
Thanks a lot for this work!
